modules/support.lua

```diff
@@ -0,0 +1,212 @@
+local PreferredServantArray = {}
+local PreferredCraftEssenceTable = {}
+
+-- functions
+local init
+local selectSupport
+local selectFirst
+local selectPreferred
+local selectManual
+local decideSearchMethod
+local searchMethod
+local findServants
+local findCraftEssence
+local isLimitBroken
+
+init = function()
+	local function split(str)
+		local values = {}
+	
+		for value in str:gmatch("[^,]+") do
+			value = stringUtils.trim(value)
+			
+			if value:lower() ~= "any" then
+				table.insert(values, value)
+			end
+		end
+		
+		return values
+	end
+
+	-- servants
+	for _, servant in ipairs(split(Support_PreferredServants)) do
+		servant = stringUtils.trim(servant)
+		table.insert(PreferredServantArray, servant)
+	end
+	
+	-- craft essences
+	for _, craftEssence in ipairs(split(Support_PreferredCEs)) do
+		craftEssence = stringUtils.trim(craftEssence)
+		
+		local craftEssenceEntry =
+		{
+			Name = craftEssence:gsub("%" .. LimitBrokenCharacter, ""),
+			PreferLimitBroken = stringUtils.starts_with(craftEssence, LimitBrokenCharacter)
+		}
+		
+		table.insert(PreferredCraftEssenceTable, craftEssenceEntry)
+	end
+end
+
+selectSupport = function(selectionMode)
+	if ScreenRegion:exists(GeneralImagePath .. "support_screen.png") then
+		if selectionMode == "first" then
+			return selectFirst()
+			
+		elseif selectionMode == "preferred" then
+			local searchMethod = decideSearchMethod()
+			return selectPreferred(searchMethod)
+			
+		elseif selectionMode == "manual" then
+			selectManual()
+			
+		else
+			scriptExit("Invalid support selection mode: \"" + selectionMode + "\".")
+		end
+	end
+
+	return false
+end
+
+selectFirst = function()
+	click(Location(1900,500))
+	return true
+end
+
+selectPreferred = function(searchMethod)
+	local numberOfSwipes = 0
+	local numberOfUpdates = 0
+	
+	while (true)
+	do
+		local support = searchMethod()
+		if support ~= nil then
+			click(support)
+			return true
+		end
+
+		if numberOfSwipes < Support_SwapsPerRefresh then
+			swipe(Location(1200, 1150), Location(1200, 800))			
+			numberOfSwipes = numberOfSwipes + 1
+			wait(0.3)
+		elseif numberOfUpdates < Support_MaxRefreshes then		
+			click(UpdateClick)
+			wait(1)
+			click(UpdateYesClick)
+			wait(3)
+
+			numberOfUpdates = numberOfUpdates + 1
+			numberOfSwipes = 0
+		else -- not found :(
+			click(ListTopClick)
+			return selectSupport(Support_FallbackTo)
+		end
+	end
+end
+
+selectManual = function()
+	scriptExit("Support selection mode set to \"manual\".")
+end
+
+decideSearchMethod = function()
+	local hasServants = #PreferredServantArray > 0
+	local hasCraftEssences = #PreferredCraftEssenceTable > 0
+
+	if hasServants and hasCraftEssences then
+		return searchMethod.byServantAndCraftEssence
+		
+	elseif hasServants then
+		return searchMethod.byServant
+		
+	elseif hasCraftEssences then
+		return searchMethod.byCraftEssence
+		
+	else
+		scriptExit("When using \"preferred\" support selection mode, specify at least one Servant or Craft Essence.")
```

Comment:
There's a new `scriptExit` call for when users leave both `Support_PreferredServants` and `Support_PreferredCEs` set to `"Any"`, as there is no point in doing so. It's better to just use the `first` option instead.